tc/core/polyhedral/cuda/memory_promotion_heuristic.cc

```diff
@@ -323,6 +323,24 @@ bool promotionImprovesCoalescing(
   return false;
 }
 
+/*
+ * Returns the union of all mapping filters to "MappingType" in "scop".
+ */
+template <typename MappingType>
+isl::union_set collectMappingsTo(const Scop& scop) {
+  auto root = scop.scheduleRoot();
+  auto domain = scop.domain();
+  auto mappingFilters = detail::ScheduleTree::collect(
+      root, detail::ScheduleTreeType::MappingFilter);
+  mappingFilters = functional::Filter(isMappingTo<MappingType>, mappingFilters);
+  auto mapping = isl::union_set::empty(domain.get_space());
+  for (auto mf : mappingFilters) {
+    auto filterNode = mf->elemAs<detail::ScheduleTreeElemMappingFilter>();
+    mapping = mapping.unite(filterNode->filter_);
```

Comment:
Doesn't this have the same issue as #505
(which you have since solved in `extractDomainToIds`)?

Actually, this looks like duplicate code wrt `extractDomainToIds`
(the mapping is represented differently, but conceptually, it's the same operation),
but I guess we can fix that up later.

Reply:
> Doesn't this have the same issue as #505
>  (which you have since solved in extractDomainToIds)?

Good point.

> Actually, this looks like duplicate code wrt extractDomainToIds
> (the mapping is represented differently, but conceptually, it's the same operation),
> but I guess we can fix that up later.

Yes, there are increasingly many places that need mapping in both forms: mupa and union set.  It should be relatively easy to extract a standalone conversion function from mupa to union set and always return mupa.  Or introduce an opaque "mapping" object that can be viewed as both types.  Created #519